### ISSUE TITLE
Overloading __getitem__ makes type not iterable

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#4859]
**Describe the bug**

When combining `__len__` and `__getitem__` to allow unpacking of class, pyright is fine, but adding overload to `__getitem__` generates an error due to a missing `__iter__`.

**To Reproduce**

The following code is accepted by pyright and deduces `a` and `b` as `int | str`:

```python
from typing import Literal

class T:
    def __len__(self) -> int:
        return 2

    def __getitem__(self, i: Literal[0, 1]) -> int | str:
        return (42, "hello")[i]

a, b = T()
```

Adding overload to `__getitem__` (to obtain proper types in `a` and `b`), generates an error:

```python
from typing import Literal, overload

class T:
    def __len__(self) -> int:
        return 2

    @overload
    def __getitem__(self, i: Literal[0]) -> int:
        ...

    @overload
    def __getitem__(self, i: Literal[1]) -> str:
        ...

    def __getitem__(self, i: Literal[0, 1]) -> int | str:
        return (42, "hello")[i]

a, b = T()
```

Running `pyright` on this outputs:

```
...\test.py
  ...\test.py:23:8 - error: "T" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
```

**Expected behavior**

I would expect no error, and eventually proper typing of `a` and `b` (in this case, `a` and `b` are `Unknown`).

**VS Code extension or command-line**

Both, CLI version is pyright 1.1.301

**Additional Contexts**

I have not found the Python requirements for `a, b = T(()` to be valid (on `T`), seems to me that it needs either `__iter__` on `T` or `__getitem__` (`__len__` does not seem mandatory).

The purpose of this is to be able to unpack `T` and get proper typing on `a` and `b`.

**Comment by erictraut** [comment]
Adding support for `__getitem__` overloads in this case would require a bunch of extra work, and I don't think it's worth the effort. The pattern you're trying to use here is very atypical, and I recommend using a different approach.

Mypy produces the same error in this case.

We might revisit this in the future if there is evidence that it would benefit other pyright users, but for now I'm going to close this as "won't fix".

**Comment by erictraut** [comment]
Actually, pyright does already support an overloaded `__getitem__`, but you need to specify an overload that accepts an `int`. You cannot specify only `Literal` types for the index. Remember, type checkers don't actually run the code at type checking time. They don't iterate through all possible indexes. They assume that the index is an `int` and "call" the method accordingly.

```python
    @overload
    def __getitem__(self, i: int) -> int | str:
        ...
```

**Comment by Holt59** [comment]
> A workaround is to make `T` a proper `Iterator` with an `__iter__` and `__next__` method. The old-style `__getitem__`/`__len__` mechanism of iteration is a fallback in the runtime.

I now that `__iter__` works but as I mentioned in the question, my goal is to have proper typing on `a` and `b` when doing `a, b = T()`, which I don't get with `__iter__`.

> Actually, pyright does already support an overloaded `__getitem__`, but you need to specify an overload that accepts an `int`. You cannot specify only `Literal` types for the index. Remember, type checkers don't actually run the code at type checking time. They don't iterate through all possible indexes like the interpreter does. They assume that the index is an `int` and evaluate the method call accordingly.
> 
> ```python
>     @overload
>     def __getitem__(self, i: int) -> int | str:
>         ...
> ```

Unfortunately, adding this overload also deduces `a` and `b` as `int | str`.

I guess there are specific implementation in type checker to handle `tuple` or `NamedTuple` since those works fine, my question would be if there is a way to make the type checker work similarly for custom classes (without inheriting `tuple` or `NamedTuple`)?

**Comment by erictraut** [comment]
Yes, there's a bunch of special-case code to handle unpacking of (and iterating over) `tuple` and `NamedTuple` objects and their subclasses. You'd need to inherit from `tuple` to make use of that special casing.

I can't think of any way to do it with a non-tuple subclass. I thought maybe there would be a clever way to leverage [PEP 646](https://peps.python.org/pep-0646/) and `TypeVarTuple`, but I haven't come up with any way using the `Iterator` protocol or `__getitem__`.

**Comment by anabelle2001** [comment]
The original example <https://github.com/microsoft/pyright/issues/4859#issue-1645240805> stopped working in v1.1.335:

```py
from typing import Literal

class T:
    def __len__(self) -> int:
        return 2

    def __getitem__(self, i: Literal[0,1]) -> int | str:
        return (42, "hello")[i]

a, b = T()
```

```sh
$ env PYRIGHT_PYTHON_FORCE_VERSION=1.1.334 pyright ./example.py
0 errors, 0 warnings, 0 informations 
```

```sh
$ env PYRIGHT_PYTHON_FORCE_VERSION=1.1.335 pyright ./example.py

C:\Users\anabelle\example.py
  C:\Users\anabelle\example.py:10:8 - error: "T" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

IMO, pyright emitting an error is correct here. In the above example, `__len__` is allowed to return any int, whereas `__getitem__` can only accept 0 or 1.

A workaround for anyone stumbling upon this thread is to type the `__getitem__` method's parameter as `i: int`, as opposed to `i: Literal[0,1]`, as erictraut suggested above: https://github.com/microsoft/pyright/issues/4859#issuecomment-1488697659
